net: lwm2m: Add event LWM2M_RD_CLIENT_EVENT_REG_UPDATE

juhaylinen opened the pull request:
New event LWM2M_RD_CLIENT_EVENT_REG_UPDATE to indicate application that engine starts registration update.

Signed-off-by: Juha Ylinen <juha.ylinen@nordicsemi.no>

juhaylinen committed:
net: lwm2m: Add event LWM2M_RD_CLIENT_EVENT_REG_UPDATE

New event LWM2M_RD_CLIENT_EVENT_REG_UPDATE to indicate
application that engine starts registration update.

Signed-off-by: Juha Ylinen <juha.ylinen@nordicsemi.no>

rlubos reviewed (commented):
What's the use-case for the event?

Please have a look at the CI failures, they seem to be significant ones.

juhaylinen committed:
tests: lwm2m_rd_client: Update tests

Update unit tests for lwm2m_rd_client.c after adding new event
LWM2M_RD_CLIENT_EVENT_REG_UPDATE.

Signed-off-by: Juha Ylinen <juha.ylinen@nordicsemi.no>

juhaylinen commented:
@rlubos The idea is to trigger connection pre-evaluation when application receives the event. Based on the evaluation parameters, the application may postpone the registration update.

SeppoTakalo commented:
I had to take a look at the RD client..

I sounds a bit scary, but I think this will work. 
```
// somewhere
set_sm_state(ENGINE_UPDATE_REGISTRATION)
    |
    v
client.ctx->event_cb(client.ctx, LWM2M_RD_CLIENT_EVENT_REG_UPDATE);
    |
    v
/* Some connection evaluation */
int val = evaluate_my_connection();
if (val < TRESHOLD) lwm2m_rd_client_pause()
                                        |
                                        v
             client.engine_state = ENGINE_SUSPENDED;
```

So as a side-effect of callback, we end up changing the state that caused the callback.
